calc_si_align.py

```diff
@@ -0,0 +1,219 @@
+    # Attitude quaternion for the as-run PCAD attitude
+    q_pcad = Quat([-0.18142595, -0.37811633, -0.89077416, 0.17502588])
+
+    # Target coordinates and quaternion, using the PCAD roll
+    ra_targ, dec_targ = 191.321250, 27.125556
+    q_targ = Quat([ra_targ, dec_targ, q_pcad.roll])
+
+    # Offsets from OR (Target DY, DZ) in degrees
+    dy_offset, dz_offset = 0.002500, -0.004167
+
+    # Sanity check that offset between as-planned PCAD and target are as expected.  This
+    # uses the baseline ODB_SI_ALIGN.  The calc_offsets() function has been independently
+    # validated for planning purposes.
+    dy_nom_align, dz_nom_align = calc_offsets(ra_targ, dec_targ,
+                                              q_pcad.ra, q_pcad.dec, q_pcad.roll)
+    assert np.allclose(dy_nom_align, dy_offset, atol=3e-6)  # 0.01 arcsec
+    assert np.allclose(dz_nom_align, dz_offset, atol=3e-6)
+
+    # Test calc_pcad_from_targ() for baseline SI_ALIGN
+    q_pcad_calc = calc_pcad_from_targ(ra_targ, dec_targ, q_pcad.roll, dy_offset, dz_offset)
+    dq = q_pcad.inv() * q_pcad_calc
+    assert np.allclose(dq.q[0] * 2, 0, atol=36 * ARCSEC2RAD)
+    assert np.allclose(dq.q[1] * 2, 0, atol=0.01 * ARCSEC2RAD)
+    assert np.allclose(dq.q[2] * 2, 0, atol=0.01 * ARCSEC2RAD)
+
+    # Compute PCAD attitude using a new SI ALIGN values based on OR-list TARGET_OFFSET
+    # values and with no target offset applied.  This must match the as-planned
+    # attitude using the baseline ODB_SI_ALIGN and WITH a target offset applied.
+    si_align = calc_si_align(dy_offset, dz_offset, check_consistency=True)
+    q_pcad_align = q_targ * Quat(si_align)
+
+    dy_align, dz_align = calc_offsets(ra_targ, dec_targ, *q_pcad_align.equatorial)
+    dq = q_pcad.inv() * q_pcad_align
+
+    # Check delta quaternion (error) is less than 0.01 arcsec
+    # in pitch and yaw, 36 arcsec in roll.
+    assert np.allclose(dq.q[0] * 2, 0, atol=36 * ARCSEC2RAD)
+    assert np.allclose(dq.q[1] * 2, 0, atol=0.01 * ARCSEC2RAD)
+    assert np.allclose(dq.q[2] * 2, 0, atol=0.01 * ARCSEC2RAD)
+
+    # Check offsets match to better than 0.01 arcsec
+    assert np.allclose(dy_align, dy_offset, atol=3e-6)
+    assert np.allclose(dz_align, dz_offset, atol=3e-6)
+
+
+def calc_si_align(dy_align=0.0, dz_align=0.0, check_consistency=True):
+    """
+    Compute the SI alignment matrix that includes an alignment offset
+    (dy_align, dz_align) from the baseline ODB_SI_ALIGN.
+
+    :param dy_align: Aimpoint alignment offset in Dy (deg)
+    :param dz_align: Aimpoint alignment offset in Dz (deg)
+    :param check_consistency: double check the output by essentially inverting it
+
+    :rtype: 3x3 alignment matrix
+    """
+    # Define new SI_ALIGN as a delta quaternion update to existing.
+    # The minus signs here were empirically determined to pass tests.
+    # This is the only place where signs are chosen without a clear justification.
+    si_align = Quat([-dy_align, -dz_align, 0.0]) * Quat(ODB_SI_ALIGN)
```

Comment:
My mistake, flipping the signs has a different effect.  I'm going to remove my previous comment and look at which rotation I defined when I did this.